cmd/cgr-tester/callproc.go

```diff
@@ -0,0 +1,165 @@
+/*
+Real-time Online/Offline Charging System (OCS) for Telecom & ISP environments
+Copyright (C) ITsysCOM GmbH
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>
+*/
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/cenkalti/rpc2"
+	"github.com/cgrates/cgrates/sessions"
+	"github.com/cgrates/cgrates/utils"
+)
+
+var (
+	brpc             *rpc2.Client
+	disconnectEvChan = make(chan *utils.AttrDisconnectSession, 1)
+	cpsCounter       int
+)
+
+func handleDisconnectSession(clnt *rpc2.Client,
+	args *utils.AttrDisconnectSession, reply *string) error {
+	disconnectEvChan <- args
+	*reply = utils.OK
+	return nil
+}
+
+func callSession() (err error) {
+
+	var currentUsage time.Duration
+
+	st := time.Now().Add(2 * time.Second)
+	at := time.Now().Add(10 * time.Second)
+
+	event := &utils.CGREvent{
+		Tenant: *tenant,
+		ID:     "TheEventID100000",
+		Time:   utils.TimePointer(time.Now()),
+		Event: map[string]any{
+			utils.AccountField: *subject,
+			utils.Destination:  *destination,
+			utils.OriginHost:   "local",
+			utils.RequestType:  utils.MetaRated,
+			utils.SetupTime:    st,
+			utils.Source:       "cgr_tester",
+			utils.OriginID:     utils.GenUUID(),
+		},
+		APIOpts: map[string]any{},
+	}
+
+	cpsCounter += 1
+	log.Printf("current call number: %+v", cpsCounter)
+
+	if *updateInterval > *maxUsage {
+		return fmt.Errorf(`"update_interval" should be smaller than "max_usage"`)
+	} else if *maxUsage <= *minUsage {
+		return fmt.Errorf(`"min_usage" should be smaller than "max_usage"`)
+	}
+
+	tstCfg.SessionSCfg().DebitInterval = 0
+
+	clntHandlers := map[string]any{utils.SessionSv1DisconnectSession: handleDisconnectSession}
+	brpc, err = utils.NewBiJSONrpcClient(tstCfg.SessionSCfg().ListenBijson, clntHandlers)
+	if err != nil {
+		return
+	}
+
+	//
+	// SessionSv1AuthorizeEvent
+	//
+	authArgs := &sessions.V1AuthorizeArgs{
+		GetMaxUsage: true,
+		CGREvent:    event,
+	}
+	var authRply sessions.V1AuthorizeReply
+	if err = brpc.Call(utils.SessionSv1AuthorizeEvent, authArgs, &authRply); err != nil {
+		return
+	}
+	// log.Printf("auth: %+v", utils.ToJSON(authRply))
+
+	//
+	// SessionSv1InitiateSession
+	//
+	event.Event[utils.AnswerTime] = at
+	event.Event[utils.RequestType] = utils.MetaRated
+
+	initArgs := &sessions.V1InitSessionArgs{
+		InitSession: true,
+		CGREvent:    event,
+	}
+
+	var initRply sessions.V1InitSessionReply
```

Comment:
Please add a small random delay between auth and init so we can get closer to reality.